htmresearch/algorithms/image_sparse_net.py

```diff
@@ -0,0 +1,182 @@
+  """
+  Implementation of SparseNet specifically suited for natural images.
+
+  It's particularity is that a training batch is composed with patches of one
+  particular input image.
+  """
+
+  BUFFER = 4
+
+
+  def loadMatlabImages(self, path, name):
+    """
+    Loads images from a .mat file.
+    :param path:      (string)   Path to .mat file
+    :param name:      (string)   Object name in the .mat file, just before .mat
+
+    Also stores image dimensions to later the original images. If there are
+    multiple channels, self.numChannels will store the number of channels,
+    otherwise it will be set to None.
+    """
+    try:
+      images = sc.loadmat(path)[name]
+    except KeyError:
+      raise KeyError('Wrong filename for provided images.')
+
+    self._initializeDimensions(images)
+
+    return images
+
+
+  def loadNumpyImages(self, path, key=None):
+    """
+    Loads images using numpy.
+
+    :param path:      (string)   Path to data file
+    :param key:       (string)   Object key in data file if it's a dict
+
+    Also stores image dimensions to later the original images. If there are
+    multiple channels, self.numChannels will store the number of channels,
+    otherwise it will be set to None.
+    """
+    data = np.load(path)
+
+    if isinstance(data, dict):
+      if key is None:
+        raise ValueError("Images are stored as a dict, a key must be provided!")
+      try:
+        data = data[key]
+      except KeyError:
+        raise KeyError("Wrong key for provided data.")
+
+    if not isinstance(data, np.ndarray):
+      raise TypeError("Data must be stored as a dict or numpy array.")
+
+    self._initializeDimensions(data)
+
+    return data
+
+
+  def _initializeDimensions(self, inputData):
+    """
+    Stores the training images' dimensions, for convenience.
+    """
+    if len(inputData.shape) == 2:
+      self.imageHeight, self.numImages = inputData.shape
+      self.imageWidth, self.numChannels = None, None
+
+    elif len(inputData.shape) == 3:
+      self.imageHeight, \
+      self.imageWidth, \
+      self.numImages = inputData.shape
+      self.numChannels = None
+
+    elif len(inputData.shape) == 4:
+      self.imageHeight, \
+      self.imageWidth, \
+      self.numChannels, \
+      self.numImages = inputData.shape
+
+    else:
+      raise ValueError("The provided image set has more than 4 dimensions.")
+
+
+  def _getDataBatch(self, inputData):
+    """
+    Returns an array of dimensions (inputDim, batchSize), to be used as
+    batch for training data.
+
+    This implementation uses random sub-images from a random image as
+    training batch.
+
+    Images are flattened to get a 2-dimensional batch.
+    """
+    if not hasattr(self, 'numImages'):
+      self._initializeDimensions(inputData)
+
+    batch = np.zeros((self.inputDim, self.batchSize))
+    miniImageSize = int(np.sqrt(self.inputDim))
+    # choose random image
+    imageIdx = np.random.choice(range(self.numImages))
```

Comment:
Algorithmically, this first pass of the implementation is making the same choices as Bruno. That's what he does in his code. If we believe we should sample from multiple images, that's an easy change. Should I do it?